earth_enterprise/src/common/FileAccessorPluginLoader.cpp

```diff
@@ -0,0 +1,103 @@
+// Copyright 2020 The Open GEE Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "FileAccessorPluginLoader.h"
+#include "POSIXFileAccessor.h"
+#include "notify.h"
+
+void ScanPluginDirectory(const std::string &pluginDir, std::vector<std::string> &files) {
+    boost::filesystem::path p(pluginDir);
+    boost::system::error_code errorCode;
+    boost::filesystem::directory_iterator start(p, errorCode);
+    if (errorCode.value() != 0){
+        notify(NFY_WARN, "Unable to scan file accessor plugin directory, %s. %s.", pluginDir.c_str(), errorCode.message().c_str());
+        return;
+    }
+
+    boost::filesystem::directory_iterator end;
+    std::transform(start, end, std::back_inserter(files), 
+        [pluginDir](const boost::filesystem::directory_entry& entry) {
+            return pluginDir + entry.path().leaf().string(); });
+}
+
+FileAccessorPluginLoader& FileAccessorPluginLoader::Get() {
+    static FileAccessorPluginLoader loader;
+    return loader;
+}
+
+void FileAccessorPluginLoader::DefaultLoadPluginsImpl (const std::string &pluginDirectory, std::vector<PluginFactory> &factories) {
+    assert(factories.size() == 0);
+    
+    std::vector<std::string> files;
+    ScanPluginDirectory(pluginDirectory, files);
+    for (auto s : files) {
+        FileAccessorPluginLoader::PluginHandle handle = dlopen(s.c_str(), RTLD_NOW);
+        if (handle){
+            get_factory_t get_factory = reinterpret_cast<get_factory_t>(dlsym(handle, "get_factory_v1"));
+            FileAccessorFactory *pFactory = get_factory();
+            if(pFactory){
+                factories.push_back({handle,pFactory});
+            }
+            else{
+                dlclose(handle);
```

Comment:
Should we have an error message here? This would be the case where the plugin loaded successfully but it didn't implement the expected API, right?

Reply:
Yes, that's correct. I'll log an error message at that point.